src/main/java/org/jitsi/videobridge/cc/vp8/VP8FrameProjection.java

```diff
@@ -0,0 +1,468 @@
+     * @return the TL0PICIDX to use in the projection of the frame that is
+     * specified as an argument.
+     */
+    private int nextTL0PICIDX(@NotNull VP8Frame nextVP8Frame)
+    {
+        return nextVP8Frame.isTL0()
+            ? VP8Frame.nextTL0PICIDX(tl0PICIDX) : tl0PICIDX;
+    }
+
+    /**
+     * Small utility method that computes and returns the RTP timestamp to use
+     * in the projection of the frame that is specified as an argument. The
+     * specified frame is assumed to be the frame that will be sent immediately
+     * after the {@link #vp8Frame}.
+     *
+     * @param nextVP8Frame the frame that will be sent immediately after
+     * {@link #vp8Frame}
+     * @return the TL0PICIDX to use in the projection of the frame that is
+     * specified as an argument.
+     */
+    private long nextTimestamp(@NotNull VP8Frame nextVP8Frame, long nowMs)
+    {
+        long delta;
+        if (!vp8Frame.matchesSSRC(nextVP8Frame))
+        {
+            // this is a simulcast switch. The typical incremental value =
+            // 90kHz / 30 = 90,000Hz / 30 = 3000 per frame or per 33ms
+            delta = 3000 * Math.max(1, (nowMs - createdMs) / 33);
+        }
+        else
+        {
+            // compute and apply a delta
+            delta = RTPUtils.rtpTimestampDiff(
+                nextVP8Frame.getTimestamp(), vp8Frame.getTimestamp());
+        }
+
+        long nextTimestamp = timestamp + delta;
+        return nextTimestamp & RawPacket.TIMESTAMP_MASK;
+    }
+
+    /**
+     * Small utility method that computes and returns the starting sequence
+     * number to use in the projection of the frame that will be sent
+     * immediately after the {@link #vp8Frame}.
+     *
+     * @return the starting sequence number to use in the projection of the
+     * frame that will be sent immediately after the {@link #vp8Frame}.
+     */
+    private int nextStartingSequenceNumber()
+    {
+        int vp8FrameLength = RTPUtils.getSequenceNumberDelta(
+            vp8Frame.getMaxSequenceNumber(),
+            vp8Frame.getStartingSequenceNumber());
+
+        int nextStartingSequenceNumber
+            = startingSequenceNumber + vp8FrameLength + 1;
+
+        return nextStartingSequenceNumber & RawPacket.SEQUENCE_NUMBER_MASK;
+    }
+
+    /**
+     * Small utility method that computes and returns the extended picture id to
+     * use in the projection of the frame that will be sent immediately after
+     * the {@link #vp8Frame}.
+     *
+     * @return the extended picture id to use in the projection of the frame
+     * that will be sent immediately after the {@link #vp8Frame}.
+     */
+    private int nextExtendedPictureId()
+    {
+        return (extendedPictureId + 1)
+            & DePacketizer.VP8PayloadDescriptor.EXTENDED_PICTURE_ID_MASK;
+    }
+
+    /**
+     * Rewrites an RTP packet. If this is the first packet of a frame and if a
+     * packet cache has been specified, we piggy-backing any missed packets.
```

Comment:
In other words, once a frame has has been fully projected (<=> all the packets of a frame have been projected) it's removed from the map to prevent the map from growing too big. I'll add that in the comment and try to clarify this part a bit.

This is actually implemented in a very poor way and not as I was initially planning (if you look at the implementation of VP8FrameProjection#isFullyProjected() method).

Reply:
what is the scenario where we have to piggyback packets?

Reply:
The decision whether to accept/drop a frame takes place when the first packet of a frame is received. 

This is to keep things simple (i.e. make it easy to compute the starting sequence number of the projection of an accepted frame).

This, however, means that we may miss ignore some re-ordered packets of a frame that we end up accepting. In that case, we dive into the ingress cache, fetch the ignored packet and piggy-back it to the packet that we currently process.

Does that make sense? Also have a look in `VP8AdaptiveTrackProjectionContext#createVP8FrameProjection`.